### DISCUSSION TITLE
Is it possible to have the login / registration form in a modal, rather than a page?

### DISCUSSION THREAD

**Discussion by sebpowell** [nextauthjs/next-auth#791] [Help]
Is it possible to have the login / registration form in a modal, rather than a page?

My use case is as follows:

1.) I'm building a membership-based eCommerce website

2.) I want all product and category pages to be crawlable for SEO purposes

3.) If a non-authenticated user reaches a 'protected' page, they're shown an overlay to register / login. On success, the overlay disappears. 

As far as I can tell, this library only supports standalone pages, but maybe I've missed something?

Any pointers would be greatly appreciated!

**Comment by iaincollins** [comment]
Hi @sebpowell!

You can create your own [customised pages](https://next-auth.js.org/configuration/pages) for all pages returned by NextAuth.js.

You can also use the [client API](https://next-auth.js.org/getting-started/client) to create your own UI / modal.

However, there isn't a easier way to do this currently; e.g. there is no built-in modal functionality.

Having tutorials - and ultimately a built in easy to use option for this - is something I'd like to have.

**Reply by luanau** [reply]
Hi @iaincollins,
Is there a register page please?
Thanks,
Luan

**Comment by sebpowell** [comment]
@iaincollins – amazing, thanks for your help! I'll give this a go – once done, I'll share my code with you, and if you think it's any good, you can use it as an example for the docs.

**Comment by iaincollins** [comment]
That would be great thanks!

**Comment by ghost** [comment]
@sebpowell Hows the progress coming on this? I'm interested in the Modal approach as well.

**Comment by irving-caamal** [comment]
> @sebpowell Hows the progress coming on this? I'm interested in the Modal approach as well.



Hey IDK why this is still open but Of course, you can use a Modal right now, you just need to run the **signIn** method and pass your provider name.

like so:
<code>    
const loginCallback = (e) => {
        e.preventDefault();

        signIn('username-login', {
                callbackUrl: `http://localhost:3000/help`,
                username: email,
                password: password
            })};

 </code>

_Lets assume this button is on a modal_
`<ModalButton onClick={loginCallback}/>`

**Comment by Sicria** [comment]
Only downside is that you cannot capture the result from `signIn` as it forces a `window.location` with no way to prevent it. So even if you to achieve a modal login, it'll send you to another page on success or failure.

Unless someone knows of a way to prevent that which I'm missing?

**Comment by paul-vd** [comment]
Also looking for a solution here, to avoid losing the state and content inside of the current page when doing a login.

**Comment by paul-vd** [comment]
The closest solution I could come up with which feels like a very bad UX 👎 

#### 1. The modal should auto hide if there is a session
```js
<Modal visible={visible && !session} />
```
#### 2. If you open the modal and there is no session trigger a window.open
```js
useEffect(() => {
  if(visible && !session){
     window.open(window.location.origin + "/api/auth/login")
 }
},[visible, session]
```

#### 3. Handle all your authentication in a separate tab/window
This will avoid losing the current page's result, but then you need to count on the user manually directing back to the tab where they initiated the authorisation request, you might be able to just pass their email as an initial query to that page and then just trigger the signIn function with an effect? will test that out.


**update:**

it is messy but it works I created a page is just dedicated to this, but this is really not a solution, would be much better if we had an entry point to avoid callbacks on the signIn

```ts
export default function LoginEmailPage({ email }: InferGetServerSidePropsType<typeof getServerSideProps>) {
  const [session] = useSession()

  useEffect(() => {
    if (user) window.close()
    if (email && !session) {
      signIn('email', { email, callbackUrl: window.location.href + '&close=true' })
    }
  }, [email, session])

  return (
    <div>
      <h1>Sending Verification Request</h1>
      <p>{email}</p>
    </div>
  )
}

export const getServerSideProps = async (context) => {
  return {
    props: {
      email: context?.query?.e || '',
      close: context?.query?.close || false
    }
  }
}

```

**Comment by paul-vd** [comment]
@iaincollins is this is on a roadmap for the near future?

as a better workaround than my previous post, I'm using [patch-package](https://www.npmjs.com/package/patch-package) to disable [this line](https://github.com/nextauthjs/next-auth/blob/5126f4e3428f2b4b8ebdd38b10066edfdd4ea2fa/src/client/index.js#L277) when passing in a `disableCallback` argument.

```js
// signIn()

if(args.disableCallback !== true){
  window.location = data.url ? data.url : callbackUrl;
}
```

Here's the patch for anyone else wanting to use this.
```patch
// next-auth+3.1.0.patch

diff --git a/node_modules/next-auth/dist/client/index.js b/node_modules/next-auth/dist/client/index.js
index 773ec1e..eb8daf9 100644
--- a/node_modules/next-auth/dist/client/index.js
+++ b/node_modules/next-auth/dist/client/index.js
@@ -296,7 +296,9 @@ var signIn = function () {
       };
       var res = yield fetch(signInUrl, fetchOptions);
       var data = yield res.json();
-      window.location = data.url ? data.url : callbackUrl;
+      if(args.disableCallback !== true){
+        window.location = data.url ? data.url : callbackUrl;
+      }
     }
   });
 
```

**Comment by paul-vd** [comment]
Something [like this](https://github.com/magiclabs/magic-js/tree/master/packages/web/src) would be the correct way to go.

**Reply by cha2hyun** [reply]
> Something [like this](https://github.com/magiclabs/magic-js/tree/master/packages/web/src) would be the correct way to go.

Link says 404 🥲

**Comment by alexvais** [comment]
Do someone could achive this?

**Comment by MohamedSaber19** [comment]
For all who are still looking for a solution for this, I achieved it by using the **signIn** method from next-auth v4 like below with **redirect** option set to false to prevent redirecting and losing the state.
find more details [here](https://next-auth.js.org/getting-started/client#using-the-redirect-false-option)

```js
  const handleLogin = async (json) => {
    const res = await signIn("credentials", {
      email: json.email,
      password: json.password,
      redirect: false,
    });
   ...
  };

**Reply by jsjoeio** [reply]
> The redirect option is only available for credentials and email providers.

Darn. Was hoping to be able to use this with an OAuth provider like Google.

**Reply by jsjoeio** [reply]
Found an example! https://github.com/arye321/nextauth-google-popup-login

**Reply by jsjoeio** [reply]
And it works beautifully! Demo:

https://user-images.githubusercontent.com/3806031/226216219-9adba368-eab4-48cc-97ee-df5a18037bbd.mp4

**Reply by guilhermeavanci** [reply]
@jsjoeio Works like a charm. Thanks!
